listenbrainz/webserver/static/js/src/BrainzPlayer.tsx

```diff
@@ -0,0 +1,403 @@
+import * as React from "react";
+import { isEqual as _isEqual, get as _get } from "lodash";
+import * as _ from "lodash";
+import PlaybackControls from "./PlaybackControls";
+import APIService from "./APIService";
+import SpotifyPlayer from "./SpotifyPlayer";
+import YoutubePlayer from "./YoutubePlayer";
+
+export type DataSourceType = {
+  playListen: (listen: Listen) => void;
+  togglePlay: () => void;
+  seekToPositionMs: (msTimecode: number) => void;
+};
+
+export type DataSourceTypes = SpotifyPlayer | YoutubePlayer;
+
+export type DataSourceProps = {
+  show: boolean;
+  playerPaused: boolean;
+  onPlayerPausedChange: (paused: boolean) => void;
+  onProgressChange: (progressMs: number) => void;
+  onDurationChange: (durationMs: number) => void;
+  onTrackInfoChange: (title: string, artist?: string) => void;
+  onTrackEnd: () => void;
+  onTrackNotFound: () => void;
+  handleError: (error: string | Error, title?: string) => void;
+  handleWarning: (message: string | JSX.Element, title?: string) => void;
+  handleSuccess: (message: string | JSX.Element, title?: string) => void;
+  onInvalidateDataSource: (
+    dataSource?: DataSourceTypes,
+    message?: string | JSX.Element
+  ) => void;
+};
+
+type BrainzPlayerProps = {
+  spotifyUser: SpotifyUser;
+  direction: BrainzPlayDirection;
+  onCurrentListenChange: (listen: Listen) => void;
+  currentListen?: Listen;
+  listens: Array<Listen>;
+  newAlert: (
+    alertType: AlertType,
+    title: string,
+    message: string | JSX.Element
+  ) => void;
+  apiService: APIService;
+};
+
+type BrainzPlayerState = {
+  currentDataSourceIndex: number;
+  currentTrackName: string;
+  currentTrackArtist?: string;
+  direction: BrainzPlayDirection;
+  playerPaused: boolean;
+  durationMs: number;
+  progressMs: number;
+  updateTime: number;
+};
+
+export default class BrainzPlayer extends React.Component<
+  BrainzPlayerProps,
+  BrainzPlayerState
+> {
+  spotifyPlayer?: React.RefObject<SpotifyPlayer>;
+  youtubePlayer?: React.RefObject<YoutubePlayer>;
+  dataSources: Array<React.RefObject<DataSourceTypes>> = [];
+
+  firstRun: boolean = true;
```

Comment:
A comment here describing why this is needed would be very helpful. Otherwise, we forget. :D